Build fails with gcc-15

Found here https://github.com/ECP-copa/Cabana/pull/804#issuecomment-2637715568:
```
#26 22.37 In file included from /home/kokkos/src/hypre/src/distributed_ls/pilut/DistributedMatrixPilutSolver.h:115,
#26 22.37                  from /home/kokkos/src/hypre/src/distributed_ls/pilut/comm.c:25:
#26 22.37 /home/kokkos/src/hypre/src/distributed_ls/pilut/./struct.h:43:26: error: 'bool' cannot be defined via 'typedef'
#26 22.37    43 |    typedef unsigned char bool;
#26 22.37       |                          ^~~~
#26 22.37 /home/kokkos/src/hypre/src/distributed_ls/pilut/./struct.h:43:26: note: 'bool' is a keyword with '-std=c23' onwards
#26 22.37 /home/kokkos/src/hypre/src/distributed_ls/pilut/./struct.h:43:4: warning: useless type name in empty declaration
#26 22.37    43 |    typedef unsigned char bool;
#26 22.37       |    ^~~~~~~
#26 22.37 gmake[2]: *** [CMakeFiles/HYPRE.dir/build.make:3901: CMakeFiles/HYPRE.dir/distributed_ls/pilut/comm.c.o] Error 1
#26 22.37 gmake[2]: *** Waiting for unfinished jobs....
#26 22.47 gmake[1]: *** [CMakeFiles/Makefile2:542: CMakeFiles/HYPRE.dir/all] Error 2
#26 22.47 gmake: *** [Makefile:136: all] Error 2
```

Should be pretty easy to reproduce in Fedora 43 (Rawhide) or in our docker container `ghcr.io/ecp-copa/ci-containers/fedora:rawhide`.

--- a/src/distributed_ls/pilut/struct.h
+++ b/src/distributed_ls/pilut/struct.h
@@ -30,21 +30,28 @@
  * $Id$
  */
 
-#ifndef __cplusplus
-#ifndef true
-# define true  1
-# define false 0
-#endif
-
+#ifndef __cplusplus  /* In C++, bool is built-in, no need to define it */
+#if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 199901L && __STDC_VERSION__ < 202311L
+/* C99 - C11: Use stdbool.h */
+#include <stdbool.h>
+#elif defined(__STDC_VERSION__) && __STDC_VERSION__ < 199901L
+/* Pre-C99: Define bool manually */
 #ifndef bool
-# ifdef Boolean
+#ifdef Boolean
    typedef Boolean bool;
-# else
+#else
    typedef unsigned char bool;
-# endif
 #endif
 #endif
- 
+#ifndef true
+#define true 1
+#endif
+#ifndef false
+#define false 0
+#endif
+#endif
+#endif /* ifndef __cplusplus */
+
 /*************************************************************************
 * This data structure holds the data distribution
 **************************************************************************/
@@ -142,7 +149,7 @@ typedef struct reducematdef ReduceMatType;
 
 
 /*************************************************************************
-* This data structure stores information about the send in each phase 
+* This data structure stores information about the send in each phase
 * of parallel hypre_ILUT
 **************************************************************************/
 struct comminfodef {

